generator/regions_kv.cpp

```diff
@@ -0,0 +1,499 @@
+  }
+
+  uint8_t GetRank() const
+  {
+    if (m_regionData.m_adminLevel >= 2 && m_regionData.m_adminLevel <= 4)
+      return m_regionData.m_adminLevel;
+
+    switch (m_regionData.m_place)
+    {
+    case PlaceType::City: return 10;
+    case PlaceType::Town: return 11;
+    case PlaceType::Village: return 13;
+    default:
+      break;
+    }
+
+    if (m_regionData.m_adminLevel >= 5 && m_regionData.m_adminLevel <= 8)
+      return m_regionData.m_adminLevel;
+
+    switch (m_regionData.m_place)
+    {
+    case PlaceType::Suburb: return 14;
+    case PlaceType::Neighbourhood: return 15;
+    case PlaceType::Hamlet: return 17;
+    case PlaceType::Locality: return 19;
+    case PlaceType::IsolatedDwelling: return 20;
+    default:
+      break;
+    }
+
+    return kNoRank;
+  }
+
+  std::string GetLabel() const
+  {
+    switch (m_regionData.m_adminLevel)
+    {
+    case 2: return "country";
+    case 4: return "region";
+    default:
+      break;
+    }
+
+    switch (m_regionData.m_place)
+    {
+    case PlaceType::City:
+    case PlaceType::Town:
+    case PlaceType::Village:
+    case PlaceType::Hamlet:
+      return "locality";
+    default:
+      break;
+    }
+
+    switch (m_regionData.m_adminLevel)
+    {
+    case 6: return "subregion";
+    default:
+      break;
+    }
+
+    switch (m_regionData.m_place)
+    {
+    case PlaceType::Suburb:
+    case PlaceType::Neighbourhood:
+      return "suburb";
+    case PlaceType::Locality:
+    case PlaceType::IsolatedDwelling:
+      return "sublocality";
+    default:
+      break;
+    }
+
+    return "";
+  }
+
+  Point Center() const
+  {
+    return m_rect.Center();
+  }
+
+  double GetArea() const
+  {
+    return m_area;
+  }
+
+  m2::RectD const & GetRect() const
+  {
+    return m_rect;
+  }
+
+private:
+  StringUtf8Multilang m_name;
+  RegionData m_regionData;
+  BoostPolygon m_polygon;
+  m2::RectD m_rect;
+  double m_area;
+};
+
+struct Node
+{
+  using Ptr = std::shared_ptr<Node>;
+  using WeakPtr = std::weak_ptr<Node>;
+  using PtrList = std::vector<Ptr>;
+
+  Node(Region const & region) : m_region(region) {}
+
+  void AddChild(Ptr child)
+  {
+    m_children.push_back(child);
+  }
+
+  void SetParent(Ptr parent)
+  {
+    m_parent = parent;
+  }
+
+  Ptr GetParent()
+  {
+    return m_parent.lock();
+  }
+
+  const PtrList & GetChildren()
+  {
+    return m_children;
+  }
+
+  bool HasChildren()
+  {
+    return m_children.size();
+  }
+
+  Region & GetData()
+  {
+    return m_region;
+  }
+
+private:
+  Region m_region;
+  PtrList m_children;
+  WeakPtr m_parent;
+};
+
+// This function is for debugging only and can be a collection of statistics.
+size_t TreeSize(Node::Ptr node)
+{
+  if (!node)
+    return 0;
+
+  if (!node->HasChildren())
+    return 1;
+
+  size_t i = 0;
+  for (auto const & n : node->GetChildren())
+    i += TreeSize(n);
+
+  return i + 1;
+}
+
+// This function is for debugging only and can be a collection of statistics.
```

Comment:
can be used for statistics collection
?

Reply:
++